File: src/source/github.zig
const std = @import("std");
const types = @import("../types.zig");
const http = @import("../util/http.zig");

pub const GithubSource = struct {
    allocator: std.mem.Allocator,
    repo: []const u8,

    pub fn init(allocator: std.mem.Allocator, repo: []const u8) !GithubSource {
        return .{ .allocator = allocator, .repo = try allocator.dupe(u8, repo) };
    }

    pub fn deinit(self: *GithubSource) void {
        self.allocator.free(self.repo);
    }

    pub fn resolve(self: GithubSource, a: std.mem.Allocator, _: []const u8) ![]const u8 {
        _ = a;
        return self.repo;
    }

    pub fn fetch(self: GithubSource, a: std.mem.Allocator, id: types.PackageIdentity) ![]u8 {
        var ver_buf: [64]u8 = undefined;
        const ver_str = try std.fmt.bufPrint(&ver_buf, "{}.{}.{}", .{ id.version.major, id.version.minor, id.version.patch });

        const url = try std.fmt.allocPrint(a, "https://api.github.com/repos/{s}/tarball/v{s}", .{ self.repo, ver_str });
        defer a.free(url);

        var client = http.HttpClient.init(a);
        defer client.deinit();

        var resp = try client.get(url);
        const body = resp.body;
        resp.body = &.{};
        return body;
    }
};

File: src/source/github.rs
use crate::source::SourceError;
use crate::types::PackageIdentity;
use crate::util::http::HttpClient;

pub struct GithubSource {
    pub repo: String,
}

impl GithubSource {
    #[must_use]
    pub const fn new(repo: String) -> Self {
        Self { repo }
    }

    #[allow(clippy::unnecessary_wraps)]
    pub async fn resolve(&self, _name: &str) -> Result<String, SourceError> {
        Ok(self.repo.clone())
    }

    pub async fn fetch(&self, identity: &PackageIdentity) -> Result<Vec<u8>, SourceError> {
        let client = HttpClient::new().map_err(|e| SourceError::NetworkError(e.to_string()))?;
        let ref_str = identity.requested_ref.as_deref().unwrap_or("HEAD");
        let url = format!(
            "https://api.github.com/repos/{repo}/tarball/{ref_str}",
            repo = self.repo
        );
        let body = client
            .get(&url)
            .await
            .map_err(|e| SourceError::NetworkError(e.to_string()))?;
        Ok(body)
    }
}

#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
    use super::*;

    #[tokio::test]
    async fn test_resolve_returns_repo() {
        let src = GithubSource::new("owner/repo".to_string());
        assert_eq!(src.resolve("any").await.unwrap(), "owner/repo");
    }

    #[test]
    fn test_github_source_new() {
        let src = GithubSource::new("user/project".to_string());
        assert_eq!(src.repo, "user/project");
    }
}
